pysb/simulator/base.py

```diff
@@ -735,15 +783,230 @@ def observables(self):
+        (optional); and the following attributes:
+        simulator_class (pickled Class), simulator_kwargs (pickled dict),
+        squeeze (bool), simulations_per_param_set (int), pysb_version (str),
+        creation_date (ISO 8601 format).
+
+        Parameters
+        ----------
+        filename: str
+            Filename to which the data will be saved
+        dataset_name: str or None
+            Dataset name. If None, will default to the name of the simulator
+            class. If the dataset_name already exists within the group,
+            a ValueError is raised.
+        group_name: str or None
+            Group name. If None, will default to the name of the model.
+        append: bool
+            If False, raise IOError if the specified file already exists. If
+            True, append to existing file (or create if it doesn't exist).
+        include_obs_exprs: bool
+            Whether to save observables and expressions in the file or not.
+            If they are not included, they can be recreated from the model
+            and species trajectories when loaded back into PySB, but you may
+            wish to include them for use with external software, or if you
+            have complex expressions which take a long time to compute.
+        """
+        if h5py is None:
+            raise Exception('Please install the h5py package for this feature')
+
+        if self._y is None and not include_obs_exprs:
+            warn('This SimulationResult has no trajectories - '
+                 'you will need to set include_obs_exprs=True if '
+                 'you wish to save observables and expressions')
+
+        if group_name is None:
+            group_name = self._model.name
+        if dataset_name is None:
+            dataset_name = self.simulator_class.__name__
+
+        # As per h5py docs, we should encode strings using numpy.string_
+        # http://docs.h5py.org/en/latest/strings.html
+        enpickle = lambda obj: np.string_(pickle.dumps(obj))
+
+        with h5py.File(filename, 'a' if append else 'w-') as hdf:
+            # Get or create the group
+            try:
+                grp = hdf.create_group(group_name)
+                grp.create_dataset('_model', data=enpickle(self._model))
+            except ValueError:
+                grp = hdf[group_name]
+                model = pickle.loads(grp['_model'][()])
+                if model.name != self._model.name:
+                    raise ValueError('SimulationResult model has name "{}", '
+                                     'but the model in HDF5 file group "{}" '
+                                     'has name "{}"'.format(self._model.name,
+                                                            group_name,
+                                                            model.name))
+
+            # Create the result dataset, which is actually a nested HDF group
+            dset = grp.create_group(dataset_name)
```

Comment:
I had the idea that it might be helpful to allow user-specified attributes on this group, to store notes about each simulation run. Nothing fancy, just numbers and strings I think. Maybe it's an explicit argument to `save` or maybe it's set as a property on `self` and pulled from there.

Reply:
Good idea. I implemented this as a `custom_attrs` dict on `SimulationResult`, where the values get pickled upon saving to HDF5.

Reply:
Well, actual HDF5 attributes would be more helpful, so the values are accessible from tools other than pysb (or at least other than Python). That's why I suggested restricting the values, so they can be stored directly as attributes. The keys in `custom_attrs` would of course also be restricted to strings.

I realize now that some sort of namespacing would be needed to be able to store each element in its own attribute. Could either keep them at this level by prefixing the keys, or add a new empty child group just to hang the attributes on.

Reply:
The `custom_attrs` dictionary entries are already stored separately as HDF5 attributes, as you suggest. Each key in `custom_attrs` is prefixed with "usrattr_" when saved to HDF5. Each value is pickled to allow storing arbitrary data types, but the strings are still human readable. E.g. the pickled string `NFsim without expressions` is stored as `S'NFsim without expressions'\np0\n.`. Would be fairly easy for non-Python software to extract the string, and the added flexibility of being able to store multiple object types seems worth having, no?